lib/base64.h

```diff
@@ -32,71 +32,112 @@
+                *p++ = sEncodingTable[((data[i] & 0x3) << 4) | ((int) (data[i + 1] & 0xF0) >> 4)];
+                *p++ = sEncodingTable[((data[i + 1] & 0xF) << 2) | ((int) (data[i + 2] & 0xC0) >> 6)];
                 *p++ = sEncodingTable[data[i + 2] & 0x3F];
             }
             if (i < in_len) {
                 *p++ = sEncodingTable[(data[i] >> 2) & 0x3F];
                 if (i == (in_len - 1)) {
                     *p++ = sEncodingTable[((data[i] & 0x3) << 4)];
                     *p++ = '=';
-                }
-                else {
-                    *p++ = sEncodingTable[((data[i] & 0x3) << 4) | ((int)(data[i + 1] & 0xF0) >> 4)];
+                } else {
+                    *p++ = sEncodingTable[((data[i] & 0x3) << 4) | ((int) (data[i + 1] & 0xF0) >> 4)];
                     *p++ = sEncodingTable[((data[i + 1] & 0xF) << 2)];
                 }
                 *p++ = '=';
             }
 
+            //remove '=' as Base 64 URL doesn't require padding
+            //https://www.rfc-editor.org/rfc/rfc4648#section-5
+            //and '=' is a reserved char in URL spec
+            if (url)
+            {
+                for(int i =0; i < 3; i++)
+                {
+                    if(ret.at(ret.length() - 1) == '=')
+                    {
+                        ret.pop_back();
+                    }
+                }
+            }
+
             return ret;
         }
 
-        static void Decode(const std::string& input, std::vector<uint8_t>& out) {
-            static constexpr unsigned char kDecodingTable[] = {
-              64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-              64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-              64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 62, 64, 64, 64, 63,
-              52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 64, 64, 64, 64, 64, 64,
-              64,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14,
-              15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 64, 64, 64, 64, 64,
-              64, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40,
-              41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 64, 64, 64, 64, 64,
-              64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-              64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-              64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-              64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-              64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-              64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-              64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-              64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64
-            };
-
+        static void Decode(const std::string& input, const unsigned char* kDecodingTable, std::vector<uint8_t>& out) {
```

Comment:
we can return the out vector by value, it's efficient in c++11

Reply:
i'll leave this as is to minimise the changes